core/system/src/ipc.c

```diff
@@ -214,6 +214,15 @@ void ipc_drain_queue(void)
         return;
     }
 
+    if (!uvisor_pool_queue_is_sane(send_queue,
+                                   send_queue->pool,
+                                   send_array,
+                                   sizeof(*send_array),
+                                   UVISOR_IPC_SEND_SLOTS)) {
+        /* This shouldn't happen in a non-malicious box. */
```

Comment:
This can happen in a non-malicious box if the spinlock is held by user code. Consider locking around the outside of the sanity check instead of inside the sanity check, so you can tell the difference between insane data structures and not able to get the lock.

Reply:
Locking outside would make the sanity check API exported incomplete. We could use different return values to indicate whether the data structure is corrupted or we cannot lock.

Reply:
Changing the return value is also acceptable.

Reply:
Any guidance on error code?

Reply:
Add something in https://github.com/ARMmbed/uvisor/blob/master/api/inc/halt_exports.h, so we can use it from test code as well.